File: p_053.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int max_sum = nums[0];
        int cur_sum = 0;
        
        for (int i = 0; i < nums.size(); i++) {
            cur_sum += nums[i];
            max_sum = max(max_sum, cur_sum);
            cur_sum = max(cur_sum, 0);
        }
        
        return max_sum;
    }
};

File: python/p_053.py
class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def maxSubArray(self, nums):
        max_sum = 0
        current_sum = 0
        
        for i in xrange(len(nums)):
            current_sum += nums[i]
            if max_sum < current_sum:
                max_sum = current_sum
            if current_sum < 0:
                current_sum = 0
        if max_sum == 0:
            max_sum = max(nums)

        return max_sum
        
